src/nb_enum.cpp

```diff
@@ -17,12 +17,21 @@ using enum_map = tsl::robin_map<int64_t, int64_t, int64_hash>;
 
 PyObject *enum_create(enum_init_data *ed) noexcept {
     // Update hash table that maps from std::type_info to Python type
-    auto [it, success] = internals->type_c2p_slow.try_emplace(ed->type, nullptr);
-    if (!success) {
-        PyErr_WarnFormat(PyExc_RuntimeWarning, 1, "nanobind: type '%s' was already registered!\n", ed->name);
-        PyObject *tp = (PyObject *) it->second->type_py;
-        Py_INCREF(tp);
-        return tp;
+    nb_internals *internals_ = internals;
+    bool success;
+    nb_type_map_slow::iterator it;
+
+    {
+        lock_internals guard(internals_);
+        std::tie(it, success) = internals->type_c2p_slow.try_emplace(ed->type, nullptr);
```

Comment:
Fixed. The copy is not needed for functionality/safety. The reason I put it is minor: accessing a global variable in a shared library requires an indirect lookup through the global offset table. Furthermore, in code with function calls and multiple compilation units,  it can be difficult for the compiler to prove that the global isn't overwritten in the meantime, requiring a redundant repeated load operation.